File: 2024/09/python/main.py
from fileinput import input

disk = list(input())[0].strip()

id_ = -1
blocks = []
tail = 0
file = True
for c in disk:
    blocks.append([tail, tail + int(c), (id_ := id_ + 1) if file else -1])
    tail += int(c)
    file = not file


expand_blocks = []
for a, b, x in sorted(blocks):
    for _ in range(b - a):
        expand_blocks.append(x)


def debug(xs):
    for a, b, x in sorted(xs):
        if x == -1:
            x = "."
        print(str(x) * (b - a), end="")
    print()


# Silver
tail = len(expand_blocks) - 1
fill = expand_blocks.index(-1)
while fill < tail:
    expand_blocks[fill] = expand_blocks[tail]
    expand_blocks[tail] = -1
    while expand_blocks[tail] == -1:
        tail -= 1
    while expand_blocks[fill] != -1:
        fill += 1

silver = sum(idx * n for idx, n in enumerate(expand_blocks) if n != -1)

print(blocks)

gold = 0

print("silver:", silver)
print("gold:", gold)

File: 2024/09/rust/src/main.rs
#![feature(iter_array_chunks)]

use std::io;

fn main() -> io::Result<()> {
    let disk_map = io::stdin()
        .lines()
        .flatten()
        .next()
        .unwrap()
        .chars()
        .map(|c| c as u8 - b'0')
        .collect();

    let silver: usize = silver(&disk_map);
    let gold: u64 = 0;
    println!("silver: {silver}");
    println!("gold: {gold}");

    return Ok(());
}

fn silver(input: &Vec<u8>) -> usize {
    let mut blocks = input
        .iter()
        .chain([0, 0].iter()) // lol
        .array_chunks()
        .enumerate()
        .map(|(n, [&f, &s])| [[Some(n)].repeat(f as usize), [None].repeat(s as usize)])
        .flatten()
        .flatten()
        .collect::<Vec<_>>();

    let mut l = 0;
    let mut r = blocks.len() - 1;
    loop  {
        while blocks[r].is_none() {
            r -= 1;
        }
        while blocks[l].is_some() {
            l += 1;
        }
        if l >= r {
            break;
        }
        let [left, right] = blocks.get_disjoint_mut([l, r]).unwrap();
        left.replace(right.take().unwrap());
    }

    blocks.iter().enumerate().map(|(n, id)| n * id.unwrap_or_default()).sum()
}
